pkg/distributor/query.go

```diff
@@ -196,15 +199,33 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 		resp.Chunkseries = append(resp.Chunkseries, series)
 	}
 	for _, series := range hashToTimeSeries {
-		sort.Sort(byTimestamp(series.Samples))
 		resp.Timeseries = append(resp.Timeseries, series)
 	}
 
 	return resp, nil
 }
 
-type byTimestamp []client.Sample
+// Merges and dedupes two sorted slices with samples together, storing result in the first slice.
+// If a and b overlap, weird things will happen.
+func mergeSamplesIntoFirst(a, b []ingester_client.Sample) []ingester_client.Sample {
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].TimestampMs < b[j].TimestampMs {
+			i++
+		} else if a[i].TimestampMs > b[j].TimestampMs {
+			a = append(a, ingester_client.Sample{}) // make sure 'a' has enough space
+			copy(a[i+1:], a[i:])                    // shift elements in 'a'
```

Comment:
If `b` in entirely smaller than `a`, then there is a lot of copying going on. How do we want to deal with that? How expensive is this copying?

Reply:
Primary reason to use existing slice was that most typical scenario is merging two slices with same content – an extra allocation seemed pointless. We can also be smart about it, and simply check if they are really the same, and if not, do the standard merge. With no extra copying inside the slice [which can arguably make some elements to be copied too many times].